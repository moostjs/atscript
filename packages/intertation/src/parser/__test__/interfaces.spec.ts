import { describe, expect, it } from 'vitest'

import { parseItn } from '..'

describe('interfaces', () => {
  it('simple interface', () => {
    const result = parseItn(`interface IName { a: string; b: number }`)
    expect(result.toString()).toMatchInlineSnapshot(`
      "● [interface] "IName" type: interface <identifier>
        = [structure] "{"  (
            ● [prop] "a": [ref] "string" <>
            ● [prop] "b": [ref] "number"
          )
      "
    `)
    expect(result.messages).toHaveLength(0)
  })
  it('export interface', () => {
    const result = parseItn(`export interface IName { a: string; b: number }`)
    expect(result.toString()).toMatchInlineSnapshot(`
      "● [interface] "IName" export: export <identifier> type: interface <identifier>
        = [structure] "{"  (
            ● [prop] "a": [ref] "string" <>
            ● [prop] "b": [ref] "number"
          )
      "
    `)
    expect(result.messages).toHaveLength(0)
  })
  it('interface with optional prop', () => {
    const result = parseItn(`interface IName { a: string; b?: number }`)
    expect(result.toString()).toMatchInlineSnapshot(`
      "● [interface] "IName" type: interface <identifier>
        = [structure] "{"  (
            ● [prop] "a": [ref] "string" <>
            ● [prop] "b" optional: ? <punctuation>: [ref] "number"
          )
      "
    `)
    expect(result.messages).toHaveLength(0)
  })
  it('interface with complex prop type', () => {
    const result = parseItn(`interface IName {
        a: string | number
        b: { c: string, d: number }
        c: [string, number]
        d: (string | number)[]
        e: string[]
        f: string | number | { g: string }
        g: { h: string, i: number } | { j: string, k: number }
        l: ({ m: string, n: number } | { o: string, p: number } & (string | number))[]
      }`)
    expect(result.toString()).toMatchSnapshot()
    expect(result.messages).toHaveLength(0)
  })
})

describe('referred identifiers in interfaces', () => {
  it('must detect referred identifiers', () => {
    const { nodes } = parseItn(`interface IName {prop: string}`)
    expect(nodes[0]?.referredIdentifiers.map(t => t.text)).toEqual(['string'])
  })
  it('must detect complex referred identifiers', () => {
    const { nodes } = parseItn(
      `interface IName {prop: string, prop2: number | string, prop3: { a: Ref1; b: Ref2 }}`
    )
    expect(nodes[0]?.referredIdentifiers.map(t => t.text)).toEqual([
      'string',
      'number',
      'string',
      'Ref1',
      'Ref2',
    ])
  })
})
