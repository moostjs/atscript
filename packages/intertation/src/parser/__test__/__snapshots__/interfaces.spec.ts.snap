// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`interfaces > interface with complex prop type 1`] = `
"● [interface] "IName" type: interface <identifier>
  = [structure] "{"  (
      ● [prop] "a"
        = [group] ""  (
            ● [ref] "string" <|>
            ● [ref] "number"
          )
       <>
      ● [prop] "b"
        = [structure] "{"  (
            ● [prop] "c": [ref] "string" <>
            ● [prop] "d": [ref] "number"
          )
       <>
      ● [prop] "c"
        = [tuple] "["  (
            ● [ref] "string" <,>
            ● [ref] "number"
          )
       <>
      ● [prop] "d": [array] "["
        = [group] "("  (
            ● [ref] "string" <|>
            ● [ref] "number"
          )
       <>
      ● [prop] "e": [array] "[": [ref] "string" <>
      ● [prop] "f"
        = [group] ""  (
            ● [ref] "string" <|>
            ● [ref] "number" <|>
            ● [structure] "{"  (
                ● [prop] "g": [ref] "string"
              )
          )
       <>
      ● [prop] "g"
        = [group] ""  (
            ● [structure] "{"  (
                ● [prop] "h": [ref] "string" <>
                ● [prop] "i": [ref] "number"
              ) <|>
            ● [structure] "{"  (
                ● [prop] "j": [ref] "string" <>
                ● [prop] "k": [ref] "number"
              )
          )
       <>
      ● [prop] "l": [array] "["
        = [group] "("  (
            ● [structure] "{"  (
                ● [prop] "m": [ref] "string" <>
                ● [prop] "n": [ref] "number"
              ) <|>
            ● [group] ""  (
                ● [structure] "{"  (
                    ● [prop] "o": [ref] "string" <>
                    ● [prop] "p": [ref] "number"
                  ) <&>
                ● [group] "("  (
                    ● [ref] "string" <|>
                    ● [ref] "number"
                  )
              )
          )
      
    )
"
`;
